### ISSUE TITLE
Study guestwalk's 'base' solution

### ISSUE THREAD

**Issue by atkm** [atkm/avazu-ctr#12]
https://github.com/guestwalk/kaggle-avazu

**Comment by atkm** [comment]
Look at gen_data.py more carefully.

`base/run.py`:
- uses cols ['id','click','hour','banner_pos','device_id','device_ip','device_model','device_conn_type','C14','C17','C20','C21'], and add cols ['pub_id','pub_domain','pub_category','device_id_count','device_ip_count','user_count','smooth_user_hour_count','user_click_histroy'].
- `util/gen_data.py`: feature engineering
    + add count features
    + a row "is_app" iff site_id is null (== 85f751fd). Then define pub_id/domain/category := app or site var.
    + split the data into two based on "is_app".
    + user := ip & model if device_id is null (== a99f214a), else device_id.
    + do these for the training set (tr_\*) and validation set (va_\*).
- `util/parallelizer.py`: ? Looks like it's splitting a file.
- `mark1`: call train.cpp
- pickle, merge (this is where predict_proba is computed), and unpickle (write predicted proba to a csv).

`base/converter/2.py`:
- for features where counts are used, use the raw feature if its count is high enough; else use its count. The solution uses count > 1000 as a threshold for device_ip/id, and count > 30 for user and hourly user.

**Comment by atkm** [comment]
Slide 8 of https://www.csie.ntu.edu.tw/~r01922136/slides/kaggle-avazu.pdf
Implement count features for the following:

- [x] device_ip, device_id
- [x] user
- [x] hourly user
- [ ] hourly impression. They define an impression as the concatenation of all 'raw features'; not sure what they mean by that. An impression usually means that an ad is shown to an user.

Slide 10:

- [ ] click history
